hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataPortableResolver.java

```diff
@@ -162,98 +177,65 @@ private static KvMetadata resolveMetadata(
         return new KvMetadata(
                 fields,
                 GenericQueryTargetDescriptor.DEFAULT,
-                new PortableUpsertTargetDescriptor(clazz)
+                new PortableUpsertTargetDescriptor(classDefinition)
         );
     }
 
-    @Nullable
-    private static ClassDefinition findClassDefinition(
-            boolean isKey,
-            Map<String, String> options,
-            InternalSerializationService serializationService
-    ) {
-        Tuple3<Integer, Integer, Integer> settings = settings(isKey, options);
-        //noinspection ConstantConditions
-        return serializationService
-                .getPortableContext()
-                .lookupClassDefinition(settings.f0(), settings.f1(), settings.f2());
-    }
-
-    @Nonnull
+    @SuppressWarnings("ReturnCount")
     private static ClassDefinition resolveClassDefinition(
-            boolean isKey,
-            Map<String, String> options,
+            PortableId portableId,
             Collection<MappingField> fields,
-            InternalSerializationService serializationService
+            PortableContext context
     ) {
-        Tuple3<Integer, Integer, Integer> settings = settings(isKey, options);
-        //noinspection ConstantConditions
-        ClassDefinition classDefinition = serializationService
-                .getPortableContext()
-                .lookupClassDefinition(settings.f0(), settings.f1(), settings.f2());
+        ClassDefinition classDefinition = context.lookupClassDefinition(portableId);
         if (classDefinition != null) {
             return classDefinition;
         }
 
-        ClassDefinitionBuilder classDefinitionBuilder = new ClassDefinitionBuilder(settings.f0(), settings.f1(), settings.f2());
-        for (MappingField field : fields) {
-            String name = field.name();
-            QueryDataType type = field.type();
-            switch (type.getTypeFamily()) {
+        return fields.stream().reduce(new ClassDefinitionBuilder(portableId), (schema, field) -> {
```

Comment:
`reduce` is a heavy tool. IMO it `foreach` with side-effects would be simpler and you wouldn't have to care if it parallel

Reply:
Stream API is heavy; it creates lots of litter. However, it improves readability by making the code more declarative. Furthermore, `resolveClassDefinition()` is called at most the number of mappings and user-defined types in them. In this particular usage, using `reduce()`, I could get rid of `break` lines, which constitute ~1/3 of the method.

Reply:
So you (ab)used `reduce` for the side effect of being able to write 

```java
return schema.addBooleanField(field.name());
```
instead of
```java
schema.addBooleanField(field.name());
break;
```
?
If you really want single line in `switch`, you could abuse `map` instead of `foreach`:
```
        ClassDefinitionBuilder schema = new ClassDefinitionBuilder(portableId);
        fields.map((field) -> {
            switch (field.type().getTypeFamily()) {
                case BOOLEAN:
                    return schema.addBooleanField(field.name());
                case TINYINT:
                    return schema.addByteField(field.name());
                case SMALLINT:
                    return schema.addShortField(field.name());
                case INTEGER:
                    return schema.addIntField(field.name());
                case BIGINT:
                    return schema.addLongField(field.name());
                case REAL:
                    return schema.addFloatField(field.name());
                case DOUBLE:
                    return schema.addDoubleField(field.name());
                case DECIMAL:
                    return schema.addDecimalField(field.name());
                case VARCHAR:
                    return schema.addStringField(field.name());
                case TIME:
                    return schema.addTimeField(field.name());
                case DATE:
                    return schema.addDateField(field.name());
                case TIMESTAMP:
                    return schema.addTimestampField(field.name());
                case TIMESTAMP_WITH_TIME_ZONE:
                    return schema.addTimestampWithTimezoneField(field.name());
                default:
                    // validated earlier, skip whole __key & this
                    return schema;
            }
        });
        return schema.build();
```

I didn't mean performance (but that might be also something not to ignore) but time it takes to read and understand the code.  Is this common idiom in java?

Reply:
It seems you don't like the idea of "reducing a group of elements to a builder." However, Java architects disagree with you: they allow reducing objects of type X to object of type Y. Additionally, in your variant, the return value of `map()` is ignored, and you rely on the mutability of `ClassDefinitionBuilder`. I would rather call such usage an "abuse."

Reply:
> they allow reducing objects of type X to object of type Y

Yes, but `reduce` is "combinable", necessity to use `ExceptionUtil::notParallelizable` is a sign that we use wrong tool.
BTW, reduce is "not constrained to execute sequentially" (see javadoc), but it does not mean only in parallel (threads), it might as well be executed using SIMD or by combining results of subexecutions (probably no sane implementation would need that, but the contract does not forbid it).

Also from javadoc (emphasis mine):
> However, there are good reasons to prefer a reduce operation over a mutative accumulation such as the above. Not only is a reduction "more abstract" -- it operates on the stream as a whole rather than individual elements -- but a **properly constructed** reduce operation is **inherently parallelizable**


> return value of map() is ignored

yes, but in case of `reduce` it is always the same object, so the return value is also useless there. And, as I said before, `foreach` would convey the intention in the most explicit way.

> you rely on the mutability of ClassDefinitionBuilder

builders by definition are mutable

----

All the above is a matter of taste and style, but I see one real problem: I do not see any guarantee of order of invocations in contract for `reduce`, but order is critical for `ClassDefinitionBuilder`. Do we rely of implementation details of java library?

Reply:
I deleted `ExceptionUtil.notParallelizable()` and introduced an alternative to `reduce` and `collect`, which is not parallelizable. It eliminates the need for a combiner by processing elements in the order they appear in the stream.